### DISCUSSION TITLE
How to assign to a new variable the current time using time::PrimitiveDateTime type?

### DISCUSSION THREAD

**Discussion by frederikhors** [time-rs/time#507] [Question] [answered]
I'm using `time::PrimitiveDateTime` and I would like to create a new var with the current time but I cannot find how. Isn't there something like `now()`?

What about `Instant::now()`?

Example:

```rust
pub struct Player {
  updated_at: Option<PrimitiveDateTime>
}

impl Player {
    pub fn set_updated_at(&mut self) {
        let now = Instant::now();

        self.updated_at = Some(time::PrimitiveDateTime::from(now));
    }
```

But obviously this doesn't work:

```
mismatched types
expected struct `time::PrimitiveDateTime`, found struct `time::Instant` rustc E0308
```

**Comment by jhpratt** [comment, marked as answer]
To get the current time, you have to use `OffsetDateTime` — specifically `OffsetDateTime::now_utc` or `OffsetDateTime::now_local`. This is because `Instant` is opaque and `PrimitiveDateTime` has no time zone or UTC offset — something necessary to indicate a specific point in time.

**Comment by Kinrany** [comment]
What is the canonical place to read the reasons time shouldn't be stored in UTC, or that `PrimitiveDateTime` shouldn't be used for that purpose despite being mathematically equivalent?

**Reply by jhpratt** [reply]
You are free to store it as if it is UTC, but that doesn't mean `time` is aware that's what you're doing. Having no offset and having an offset of UTC are not equivalent, just as `None` that is of type `Option<u32>` is not equivalent to zero.

**Reply by Kinrany** [reply]
I guess the reason people often try to use `PrimitiveDateTime` is that it's very common to store timestamps in UTC, and in these cases storing `OffsetDateTime` is a footgun: no way to enforce at compile time that the offset is zero other than by wrapping it in a newtype. `PrimitiveDateTime` is the only obvious alternative that has the correct shape.

This made me realize an important difference between `time` and `chrono`: time zones are dynamic in `time` and static in `chrono`. We probably need both.

Edit: `chrono` does allow dynamic time zones as one of the time zone types, hmm.

**Reply by jhpratt** [reply]
I have mulled adding a `UtcDateTime`, for what it's worth. A recent internal change would make it very easy to implement the guts, and I'd just need to write the façade.

More generally, having support for an arbitrary static offset is something I hope to achieve eventually. Unfortunately at this point, implementing it would require a few nightly features to get it looking (and working) the way I want.
